src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/AttributeSnippetFormatOnTypeProvider.cs

```diff
@@ -0,0 +1,128 @@
+            {
+                throw new ArgumentNullException(nameof(position));
+            }
+
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            bool addCursorPlaceholder;
+            if (context.Options.TryGetValue(LanguageServerConstants.ExpectsCursorPlaceholderKey, out var value) && value.IsBool)
+            {
+                addCursorPlaceholder = value.Bool;
+            }
+            else
+            {
+                // Temporary:
+                // no-op if cursor placeholder isn't supported. This means the request isn't coming from VS.
+                edits = null;
+                return false;
+            }
+
+            if (!IsAtAttributeValueStart(context, position))
+            {
+                edits = null;
+                return false;
+            }
+
+            var cursorPlaceholder = addCursorPlaceholder ? LanguageServerConstants.CursorPlaceholderString : string.Empty;
+            var edit = new TextEdit()
+            {
+                NewText = $"\"{cursorPlaceholder}\"",
+                Range = new Range(position, position)
+            };
+
+            edits = new[] { edit };
+            return true;
+        }
+
+        private bool IsAtAttributeValueStart(FormattingContext context, Position position)
+        {
+            var syntaxTree = context.CodeDocument.GetSyntaxTree();
+
+            var absoluteIndex = position.GetAbsoluteIndex(context.SourceText);
+            var change = new SourceChange(absoluteIndex, 0, string.Empty);
+            var owner = syntaxTree.Root.LocateOwner(change);
+
+            if (owner?.Parent is MarkupLiteralAttributeValueSyntax)
+            {
+                // Accounts for
+                // 1. <Counter IncrementBy=|
+                // 2. <Counter IncrementBy|
+                // 3. <Counter IncrementBy=|
+                owner = owner.Parent;
+            }
+
+            if (owner is null ||
+                !(owner.Parent is MarkupTagHelperAttributeValueSyntax attributeValue) ||
+                !(owner.Parent.Parent is MarkupTagHelperAttributeSyntax attribute) ||
+                !(owner.Parent.Parent.Parent is MarkupTagHelperStartTagSyntax startTag))
+            {
+                // Incorrect taghelper tree structure
+                return false;
+            }
+
+            if (!attributeValue.Span.IsEmpty || string.IsNullOrEmpty(attribute.Name.GetContent()))
+            {
+                // Attribute value already started or attribute is empty
+                return false;
+            }
+
+            var ancestors = owner.Ancestors();
+            var (parentTag, parentIsTagHelper) = _tagHelperFactsService.GetNearestAncestorTagInfo(ancestors);
+            var stringifiedAttributes = _tagHelperFactsService.StringifyAttributes(startTag.Attributes);
+            var binding = _tagHelperFactsService.GetTagHelperBinding(
```

Comment:
@ajaybhargavb don't we capture the binding on one of the TagHelper parents?

Reply:
Hmm, I'm seeing it in the `MarkupTagHelperElement`. Thanks!

Updating...

Reply:
Updated.